ledger/src/blockstore.rs

```diff
@@ -2923,13 +2924,68 @@ impl Blockstore {
         self.last_root()
     }
 
+    pub fn lowest_cleanup_slot(&self) -> Slot {
+        *self.lowest_cleanup_slot.read().unwrap()
+    }
+
     pub fn storage_size(&self) -> Result<u64> {
         self.db.storage_size()
     }
 
     pub fn is_primary_access(&self) -> bool {
         self.db.is_primary_access()
     }
+
+    pub fn scan_and_fix_roots(&self, exit: &Arc<AtomicBool>) -> Result<()> {
+        if !self.is_root(self.last_root()) {
+            self.set_roots(&[self.last_root()])?;
+        }
+
+        let lowest_scan_slot = std::cmp::max(self.lowest_cleanup_slot(), self.lowest_slot());
```

Comment:
More curious  than 100% sure about this one, but shouldn't `self.lowest_slot()` always be larger than `self.lowest_cleanup_slot()`? Intuition is everything less than `self.lowest_cleanup_slot()` should be empty, i.e. this check in `self.lowest_slot()`

```
            if slot > 0 && meta.received > 0 {
                return slot;
            }
```

should not find any `meta` less than `self.lowest_cleanup_slot()`, and replay should not insert any new shreds into slots less than `self.lowest_cleanup_slot()` because it should be true `lowest_cleanup_slot < root` and we don't insert for slots `< root` in `verify_shred_slots()`

Always good to sure I guess, maybe a warning if this isn't true.

Reply:
@carllin , ah okay I'm thinking this would be fine just as `self.lowest_slot()`. Sound right?

Reply:
@CriesofCarrots yeah I think so, sounds right to me!